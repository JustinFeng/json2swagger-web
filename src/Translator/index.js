import React, { Component } from 'react';
import 'whatwg-fetch';
import { FormGroup, FormControl, Button, Grid, Row, Col } from 'react-bootstrap';
import './Translator.css';

const ROWS = 30;

class Translator extends Component {
  constructor(props) {
    super(props);
    this.state = { data: '', valid: false, result: '' };
  }

  validate = data => {
    try {
      JSON.parse(data);
    } catch (e) {
      return false;
    }

    return true;
  };

  onChange = e => {
    const data = e.target.value;
    this.setState({
      data,
      valid: this.validate(data)
    });
  };

  translate = () => {
    fetch('http://localhost:9292/translate', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: this.state.data })
      .then(response => response.text())
      .then(result => this.setState({ result }));
  };

  render = () =>
    <Grid className="App-grid">
      <Row className="show-grid">
        <Col md={5}>
          <FormGroup controlId="jsonTextArea">
            <FormControl
              componentClass="textarea"
              className="App-textarea"
              rows={ROWS}
              placeholder="Paste json data here..."
              value={this.state.data}
              onChange={this.onChange}
            />
          </FormGroup>
        </Col>
        <Col md={2} className="App-convert">
          <Button
            bsStyle="success"
            disabled={!this.state.valid}
            onClick={this.translate}>
            Translate
          </Button>
        </Col>
        <Col md={5}>
          <FormGroup controlId="swaggerTextArea">
            <FormControl
              componentClass="textarea"
              className="App-textarea"
              rows={ROWS}
              readOnly
              value={this.state.result}
            />
          </FormGroup>
        </Col>
      </Row>
    </Grid>
}

export default Translator;
