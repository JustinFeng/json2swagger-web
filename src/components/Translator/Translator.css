.App-grid {
  margin-top: 50px;
}

.App-textarea {
  resize: vertical;
  font-family: monospace;
  font-size: 14px;
}

.App-format {
  display: inline-block;
  width: 60%;
}

.App-format-label {
  margin-right: 10px;
}

.App-translate {
  margin-top: 170px;
}

.App-translate-button {
  margin-top: 20px;
}

.App-copy {
  display: none;
  position: absolute;
  bottom: 25px;
  right: 25px;
}

.App-copy.success {
  display: block;
}
