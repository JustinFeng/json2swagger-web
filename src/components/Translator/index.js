import React, { Component } from 'react';
import 'whatwg-fetch';
import { FormGroup, FormControl, Button, Grid, Row, Col, Glyphicon, ControlLabel } from 'react-bootstrap';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import apiUrl from '../../utils/apiUrl';
import './Translator.css';

const ROWS = 20;

class Translator extends Component {
  constructor(props) {
    super(props);
    this.state = { data: '', valid: false, result: '', loading: false, format: 'yaml' };
  }

  validate = data => {
    try {
      JSON.parse(data);
    } catch (e) {
      return false;
    }

    return true;
  };

  onChange = e => {
    const data = e.target.value;
    this.setState({
      data,
      valid: this.validate(data)
    });
  };

  onSelect = e => {
    const format = e.target.value;
    this.setState({
      format
    });
  }

  handleErrors = response => {
    this.setState({ loading: false });
    if (!response.ok) {
      throw Error();
    }

    return response;
  };

  translate = () => {
    this.setState({ loading: true });
    const headers = { 'Content-Type': 'application/json', 'Accept': `application/${this.state.format}` };

    fetch(`${apiUrl()}/translate`, { method: 'POST', headers, body: this.state.data })
      .then(this.handleErrors)
      .then(response => response.text())
      .then(result => this.setState({ result, resultStatus: 'success' }))
      .catch(() => this.setState({
        result: 'I was beaten by your json, please report a bug with the json.', resultStatus: 'error'
      }));
  };

  validateState = () => this.state.data ? this.state.valid ? 'success' : 'error' : null;

  renderConvert = () =>
    <Col md={2} className="App-convert">
      <FormGroup controlId="translate" className="App-translate">
        <ControlLabel className="App-format-label">Format:</ControlLabel>
        <FormControl componentClass="select" className="App-format" onChange={this.onSelect}>
          <option value="yaml">YAML</option>
          <option value="json">JSON</option>
        </FormControl>
        <Button
          className="App-translate-button"
          bsStyle="success"
          disabled={!this.state.valid}
          onClick={this.translate}>
          Translate
        </Button>
      </FormGroup>
    </Col>

  render = () =>
    <Grid className="App-grid">
      <Row className="show-grid">
        <Col md={5}>
          <FormGroup controlId="jsonTextArea" validationState={this.validateState()}>
            <FormControl
              componentClass="textarea"
              className="App-textarea"
              rows={ROWS}
              placeholder="Paste json data here..."
              value={this.state.data}
              onChange={this.onChange}
            />
            <FormControl.Feedback />
          </FormGroup>
        </Col>
        { this.renderConvert() }
        <Col md={5}>
          <FormGroup controlId="swaggerTextArea" validationState={this.state.resultStatus}>
            <FormControl
              componentClass="textarea"
              className="App-textarea"
              rows={ROWS}
              readOnly
              value={this.state.loading ? 'loading...' : this.state.result}
            />
            <FormControl.Feedback />
          </FormGroup>
          <CopyToClipboard text={this.state.result} className={`App-copy ${this.state.resultStatus}`}>
            <Button>
              <Glyphicon glyph="copy" />
            </Button>
          </CopyToClipboard>
        </Col>
      </Row>
    </Grid>
}

export default Translator;
