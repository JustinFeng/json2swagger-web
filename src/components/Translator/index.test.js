import React from 'react';
import { shallow } from 'enzyme';
import Translator from './index';

describe('Translator', () => {
  let translator;

  beforeEach(() => {
    translator = shallow(<Translator />);
  });

  describe('initialize', () => {
    it('renders', () => {
      expect(translator.state()).toEqual({ data: '', valid: false, result: '', loading: false });
      expect(translator).toMatchSnapshot();
    });
  });

  describe('input data', () => {
    let input;

    beforeEach(() => {
      input = translator.find('FormControl').at(0)
    });

    it('updates state when input data is invalid', () => {
      input.simulate('change', { target: { value: 'abc' }});

      expect(translator.state('data')).toEqual('abc');
      expect(translator.state('valid')).toBeFalsy();
      expect(translator).toMatchSnapshot();
    });

    it('updates state when input data is valid', () => {
      input.simulate('change', { target: { value: '{"some": "json"}' }});

      expect(translator.state('data')).toEqual('{"some": "json"}');
      expect(translator.state('valid')).toBeTruthy();
      expect(translator).toMatchSnapshot();
    });
  });

  describe('translate data', () => {
    describe('successful', () => {
      beforeEach(() => {
        fetch = jest.fn().mockReturnValue(Promise.resolve({ ok: true, text: () => 'result' }));

        translator.setState({ data: 'valid data', valid: true, result: '', loading: false });
        translator.find('Button').at(0).simulate('click');
      });

      it('calls translate api with input data', () => {
        expect(fetch.mock.calls.length).toBe(1);
        expect(fetch.mock.calls[0][0]).toEqual('http://localhost:9292/translate');
        expect(fetch.mock.calls[0][1].method).toEqual('POST');
        expect(fetch.mock.calls[0][1].body).toEqual('valid data');
      });

      it('updates state with translation result', () => {
        expect(translator.state('result')).toEqual('result');
        expect(translator.state('resultStatus')).toEqual('success');
        expect(translator).toMatchSnapshot();
      });
    });

    describe('unsuccessful', () => {
      beforeEach(() => {
        fetch = jest.fn().mockReturnValue(Promise.resolve({ ok: false, text: () => 'result' }));

        translator.setState({ data: 'valid data', valid: true, result: '', loading: false });
        translator.find('Button').at(0).simulate('click');
      });

      it('updates state with error message', () => {
        expect(translator.state('result')).toEqual('I was beaten by your json, please report a bug with the json.');
        expect(translator.state('resultStatus')).toEqual('error');
        expect(translator).toMatchSnapshot();
      });
    });
  });
});
