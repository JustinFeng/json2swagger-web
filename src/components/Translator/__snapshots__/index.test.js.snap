// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translator initialize renders 1`] = `
<Grid
  bsClass="container"
  className="App-grid"
  componentClass="div"
  fluid={false}
>
  <Row
    bsClass="row"
    className="show-grid"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={5}
    >
      <FormGroup
        bsClass="form-group"
        controlId="jsonTextArea"
        validationState={null}
      >
        <FormControl
          bsClass="form-control"
          className="App-textarea"
          componentClass="textarea"
          onChange={[Function]}
          placeholder="Paste json data here..."
          rows={20}
          value=""
        />
        <FormControlFeedback
          bsClass="form-control-feedback"
          bsRole="feedback"
        />
      </FormGroup>
    </Col>
    <Col
      bsClass="col"
      className="App-convert"
      componentClass="div"
      md={2}
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="success"
        disabled={true}
        onClick={[Function]}
      >
        Translate
      </Button>
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      md={5}
    >
      <FormGroup
        bsClass="form-group"
        controlId="swaggerTextArea"
      >
        <FormControl
          bsClass="form-control"
          className="App-textarea"
          componentClass="textarea"
          readOnly={true}
          rows={20}
          value=""
        />
      </FormGroup>
    </Col>
  </Row>
</Grid>
`;

exports[`Translator input data updates state when input data is invalid 1`] = `
<Grid
  bsClass="container"
  className="App-grid"
  componentClass="div"
  fluid={false}
>
  <Row
    bsClass="row"
    className="show-grid"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={5}
    >
      <FormGroup
        bsClass="form-group"
        controlId="jsonTextArea"
        validationState="error"
      >
        <FormControl
          bsClass="form-control"
          className="App-textarea"
          componentClass="textarea"
          onChange={[Function]}
          placeholder="Paste json data here..."
          rows={20}
          value="abc"
        />
        <FormControlFeedback
          bsClass="form-control-feedback"
          bsRole="feedback"
        />
      </FormGroup>
    </Col>
    <Col
      bsClass="col"
      className="App-convert"
      componentClass="div"
      md={2}
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="success"
        disabled={true}
        onClick={[Function]}
      >
        Translate
      </Button>
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      md={5}
    >
      <FormGroup
        bsClass="form-group"
        controlId="swaggerTextArea"
      >
        <FormControl
          bsClass="form-control"
          className="App-textarea"
          componentClass="textarea"
          readOnly={true}
          rows={20}
          value=""
        />
      </FormGroup>
    </Col>
  </Row>
</Grid>
`;

exports[`Translator input data updates state when input data is valid 1`] = `
<Grid
  bsClass="container"
  className="App-grid"
  componentClass="div"
  fluid={false}
>
  <Row
    bsClass="row"
    className="show-grid"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={5}
    >
      <FormGroup
        bsClass="form-group"
        controlId="jsonTextArea"
        validationState="success"
      >
        <FormControl
          bsClass="form-control"
          className="App-textarea"
          componentClass="textarea"
          onChange={[Function]}
          placeholder="Paste json data here..."
          rows={20}
          value="{\\"some\\": \\"json\\"}"
        />
        <FormControlFeedback
          bsClass="form-control-feedback"
          bsRole="feedback"
        />
      </FormGroup>
    </Col>
    <Col
      bsClass="col"
      className="App-convert"
      componentClass="div"
      md={2}
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="success"
        disabled={false}
        onClick={[Function]}
      >
        Translate
      </Button>
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      md={5}
    >
      <FormGroup
        bsClass="form-group"
        controlId="swaggerTextArea"
      >
        <FormControl
          bsClass="form-control"
          className="App-textarea"
          componentClass="textarea"
          readOnly={true}
          rows={20}
          value=""
        />
      </FormGroup>
    </Col>
  </Row>
</Grid>
`;

exports[`Translator translate data updates state with translation result 1`] = `
<Grid
  bsClass="container"
  className="App-grid"
  componentClass="div"
  fluid={false}
>
  <Row
    bsClass="row"
    className="show-grid"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={5}
    >
      <FormGroup
        bsClass="form-group"
        controlId="jsonTextArea"
        validationState="success"
      >
        <FormControl
          bsClass="form-control"
          className="App-textarea"
          componentClass="textarea"
          onChange={[Function]}
          placeholder="Paste json data here..."
          rows={20}
          value="valid data"
        />
        <FormControlFeedback
          bsClass="form-control-feedback"
          bsRole="feedback"
        />
      </FormGroup>
    </Col>
    <Col
      bsClass="col"
      className="App-convert"
      componentClass="div"
      md={2}
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="success"
        disabled={false}
        onClick={[Function]}
      >
        Translate
      </Button>
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      md={5}
    >
      <FormGroup
        bsClass="form-group"
        controlId="swaggerTextArea"
      >
        <FormControl
          bsClass="form-control"
          className="App-textarea"
          componentClass="textarea"
          readOnly={true}
          rows={20}
          value=""
        />
      </FormGroup>
    </Col>
  </Row>
</Grid>
`;
